### PR TITLE
Allow add torrent via API

### PR DESCRIPTION
**Pull request by nduhamel**

### CODE REVIEW

File: server/http.go [hunk 1]
```diff
@@ -209,6 +210,34 @@ func (a *App) deleteFile(w http.ResponseWriter, req *http.Request) {
...
 
+func (a *App) addTorrent(w http.ResponseWriter, r *http.Request) {
+	if !a.config.Downloader.Enabled {
+		a.logger.Warning("Downloader not available")
+		http.Error(w, "Downloader not enabled in your polochon", http.StatusServiceUnavailable)
+		return
+	}
+
+	req := new(struct{ URL string })
+	decoder := json.NewDecoder(r.Body)
```
Comment (gregdel) [inline server/http.go]: You can do this in one line

File: server/http.go [hunk 2]
```diff
@@ -209,6 +210,34 @@ func (a *App) deleteFile(w http.ResponseWriter, req *http.Request) {
...
+		a.logger.Warning(err.Error())
+		return
+	}
+	if req.URL == "" {
+		http.Error(w, "URL missing", http.StatusBadRequest)
+		return
+	}
+
+	if err := a.config.Downloader.Client.Download(req.URL, a.logger.WithField("function", "api")); err != nil {
+		a.logger.Warning(err)
```
Comment (gregdel) [inline server/http.go]: The logger is not set with "function == api" here

Comment (nduhamel) [inline server/http.go]: you're right

File: server/http.go (hunk 1 above)
Comment (nduhamel) [inline server/http.go]: like:

    json.NewDecoder(r.Body).Decode(req)

Comment (gregdel) [inline server/http.go]: yep
```go
if err := json.NewDecoder(r.Body).Decode(req); err != nil {
```

File: server/http.go [hunk 3]
```diff
@@ -209,6 +210,33 @@ func (a *App) deleteFile(w http.ResponseWriter, req *http.Request) {
...
+		return
+	}
+
+	if err := a.config.Downloader.Client.Download(req.URL, a.logger.WithField("function", "downloader")); err != nil {
+		a.logger.Warning(err)
+		if err == polochon.ErrDuplicateTorrent {
+			http.Error(w, "Torrent already added", http.StatusConflict)
+		}
+		http.Error(w, "Unkown error", http.StatusInternalServerError)
+	}
```
Comment (PouuleT) [inline server/http.go]: Nothing is returned if everything went right? Does http writes a 200OK automatically?

File: server/http.go [hunk 4]
```diff
@@ -209,6 +210,33 @@ func (a *App) deleteFile(w http.ResponseWriter, req *http.Request) {
...
+		return
+	}
+	if req.URL == "" {
+		http.Error(w, "URL missing", http.StatusBadRequest)
+		return
+	}
+
+	if err := a.config.Downloader.Client.Download(req.URL, a.logger.WithField("function", "downloader")); err != nil {
+		a.logger.Warning(err)
+		if err == polochon.ErrDuplicateTorrent {
```
Comment (gregdel) [inline server/http.go]: Still not the good logs, you do something like 
```go
log := a.logger.WithField("function", "downloader")
if err := a.config.Downloader.Client.Download(req.URL, log); err != nil {
    log.Warning(err)
```

File: server/http.go (hunk 3 above)
Comment (nduhamel) [inline server/http.go]: Yes

File: server/http.go (hunk 4 above)
Comment (nduhamel) [inline server/http.go]: Ok, but why because here the function is API and not the downloader?

Comment (gregdel) [inline server/http.go]: As you want, it was just to give more context to the error log. If you think it's not relevant, we can merge it as it is.

Comment (PouuleT) [inline server/http.go]: The more context we have, the better, even if it's something like 
```
"Error while downloading a torrent via the API: %q", err
```

By the way, I think it's useless to warn if the error is "Torrent already added"

File: server/http.go [hunk 5]
```diff
@@ -209,6 +210,35 @@ func (a *App) deleteFile(w http.ResponseWriter, req *http.Request) {
...
+		a.logger.Warning(err.Error())
+		return
+	}
+	if req.URL == "" {
+		http.Error(w, "URL missing", http.StatusBadRequest)
+		return
+	}
+
+	if err := a.config.Downloader.Client.Download(req.URL, a.logger.WithField("function", "downloader")); err != nil {
+		a.logger.Warning("Error while adding a torrent via the API: %q", err)
```
Comment (gregdel) [inline server/http.go]: As @PouuleT said, it will warn if the torrent is already added, is it the desired behaviour ?

Comment (nduhamel) [inline server/http.go]: nope

**Pull request merged**